internal/provider/restore_resource.go

```diff
@@ -0,0 +1,353 @@
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"net/http"
+	"time"
+
+	"github.com/cockroachdb/cockroach-cloud-sdk-go/v6/pkg/client"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+const (
+	createRestoreEndpointTimeout = time.Hour * 2
+)
+
+type restoreResource struct {
+	provider *provider
+}
+
+func NewRestoreResource() resource.Resource {
+	return &restoreResource{}
+}
+
+func (r *restoreResource) Schema(
+	_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "CockroachDB Cloud restore. Can be used to kick off a restore operation.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"destination_cluster_id": schema.StringAttribute{
+				Description: "The ID of the cluster where the backup will be restored.",
+				Required:    true,
+			},
+			"type": schema.StringAttribute{
+				MarkdownDescription: "The type of restore to be performed. Allowed values are: `CLUSTER` or `DATABASE` or `TABLE`.",
+				Required:            true,
+				Validators:          []validator.String{stringvalidator.OneOf("CLUSTER", "DATABASE", "TABLE")},
+			},
+			"backup_id": schema.StringAttribute{
+				Description: "The ID of the backup from which data will be restored. If backup_id is not set, the restore will use the latest available backup from source_cluster_id.",
+				Optional:    true,
+			},
+			"source_cluster_id": schema.StringAttribute{
+				Description: "The ID of the cluster containing the backup to be restored.",
+				Optional:    true,
+			},
+			"objects": schema.ListNestedAttribute{
+				Description: "The list of objects to restore. Required when type is DATABASE or TABLE.",
+				Optional:    true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"database": schema.StringAttribute{
+							MarkdownDescription: "The database portion of the name of the object to be restored. In the case of `tpcc.public.warehouse`, this would be `tpcc`.",
+							Required:            true,
+						},
+						"schema": schema.StringAttribute{
+							MarkdownDescription: "The schema portion of the name of the object to be restored. In the case of `tpcc.public.warehouse`, this would be `public`.",
+							Optional:            true,
+						},
+						"table": schema.StringAttribute{
+							MarkdownDescription: "The table portion of the name of the object to be restored. In the case of `tpcc.public.warehouse`, this would be `warehouse`.",
+							Optional:            true,
+						},
+					},
```

Comment:
This method of inputting data seem pretty painful.  I was hoping we could figure out a way to pull data directly from a backup object to populate this.   I don't know exactly what that would look like.  Do you have any ideas?